openeo/extra/job_management.py

```diff
@@ -0,0 +1,263 @@
+import collections
+import contextlib
+import datetime
+import json
+import logging
+import time
+from pathlib import Path
+from typing import Callable, Union, Dict, Optional
+
+import geopandas as gpd
+import pandas as pd
+import requests
+from openeo import Connection, BatchJob
+from openeo.rest import OpenEoApiError
+from openeo.util import deep_get
+
+
+_log = logging.getLogger(__name__)
+
+
+# Container for backend info/settings
+_Backend = collections.namedtuple("_Backend", ["get_connection", "parallel_jobs"])
+
+
+class MultiBackendJobManager:
+    """
+    Tracker for multiple jobs on multiple backends.
+
+    Usage example:
+
+        manager = MultiBackendJobManager()
+        manager.add_backend("foo", connection=openeo.connect("http://foo.test"))
+        manager.add_backend("bar", connection=openeo.connect("http://bar.test"))
+
+        jobs_df = pd.DataFrame(....)
+        output_file = Path("jobs.csv")
+        def start_job(row, connection, **kwargs):
+            ...
+
+        manager.run_jobs(df=df, start_job=start_job, output_file=output_file)
+
+    """
+
+    def __init__(self, poll_sleep: int = 60):
+        """Create a MultiBackendJobManager
+
+        :param poll_sleep: How many seconds to sleep between polls.
+        """
+        self.backends: Dict[str, _Backend] = {}
+        self.poll_sleep = poll_sleep
+
+    def add_backend(
+        self,
+        name: str,
+        connection: Union[Connection, Callable[[], Connection]],
+        parallel_jobs: int = 2,
+    ):
+        """Register a backend with a name and a Connection getter"""
+        if isinstance(connection, Connection):
+            c = connection
+            connection = lambda: c
+        assert callable(connection)
+        self.backends[name] = _Backend(
+            get_connection=connection, parallel_jobs=parallel_jobs
+        )
+
+    def _normalize_df(self, df: pd.DataFrame) -> pd.DataFrame:
+        # check for some required columns.
+        required_with_default = [
+            ("status", "not_started"),
+            ("id", None),
+            ("start_time", None),
+            ("cpu", None),
+            ("memory", None),
+            ("duration", None),
+            ("backend_name", None),
+        ]
+        new_columns = {
+            col: val for (col, val) in required_with_default if col not in df.columns
+        }
+        df = df.assign(**new_columns)
+        # Workaround for loading of geopandas "geometry" column.
+        if "geometry" in df.columns and df["geometry"].dtype.name != "geometry":
+            df["geometry"] = gpd.GeoSeries.from_wkt(df["geometry"])
```

Comment:
this is the only location where geopandas is really used. Can you try decoding wkt with:
https://shapely.readthedocs.io/en/stable/manual.html?highlight=wkt#shapely.wkt.loads
This way, we get a very similar geometry column without requiring geopandas.

Other proposal: try to import geopandas and only apply conversion if import succeeds?

Reply:
reason is that we try to have as few (hard) dependencies as possible, to have a very light installation

Reply:
Done: using shapely for converting the geometry column.